test/utils/deployment.go

```diff
@@ -244,3 +244,37 @@ func WaitForObservedDeployment(c clientset.Interface, ns, deploymentName string,
 		return c.Extensions().Deployments(ns).Get(deploymentName, metav1.GetOptions{})
 	}, desiredGeneration, 2*time.Second, 1*time.Minute)
 }
+
+// Pool until deployment status and its underlying resources reach the desired state.
+func WaitForDeploymentCompletes(c clientset.Interface, d *extensions.Deployment, logf LogfFn, pollInterval, pollTimeout time.Duration) error {
+	var (
+		deployment *extensions.Deployment
+		reason     string
+	)
+
+	err := wait.Poll(pollInterval, pollTimeout, func() (bool, error) {
+		var err error
+		deployment, err = c.Extensions().Deployments(d.Namespace).Get(d.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		// When the deployment status and its underlying resources reach the desired state, we're done
+		if deploymentutil.DeploymentComplete(deployment, &deployment.Status) {
```

Comment:
`DeploymentComplete` checks new deployment status to see if deployment is complete. One essential step is comparing deployment's `generation` with its `status.observedGeneration`. You cannot keep getting the latest deployment, because `generation` might be different. 

`generation` is a specific generation of the desired state, and users/clients can use `status.observedGeneration` to see if the controller has observed a specific `generation` (if `status.observedGeneration` >= a specific generation).

Reply:
Interesting. I wasn't aware that `generation` might be different. What i will do is to get the deployment only once - before polling.

Reply:
@janetkuo Do you mean something like this?
```
    deployment, err := c.ExtensionsV1beta1().Deployments(d.Namespace).Get(d.Name, metav1.GetOptions{})
	if err != nil {
		......
	}

	var reason string
	err = wait.PollImmediate(pollInterval, pollTimeout, func() (bool, error) {
		// When the deployment status and its underlying resources reach the desired state, we're done
		if deploymentutil.DeploymentComplete(deployment, &deployment.Status) {
			return true, nil
		}
		......
	})
```
The polling above will not be successful. Although it ensures validity of `observedGeneration`, the `Replicas`, `UpdatedReplicas` and `ReadyReplicas` will always be zero as we do not fetch the deployment every time (the deployment still has stale data). I think this is also why `WaitForDeploymentStatusValid` gets a new deployment for every poll.

What do you mean with `cannot keep getting the latest deployment`?

Reply:
IMO, we can save `ObservedGeneration` value and assign it to newly fetched deployment for every poll.

Reply:
You can just store the original one as the first argument of `DeploymentComplete` and pass the up-to-date deployment status as the second argument.